proto/control-api/src/callback.rs

```diff
@@ -0,0 +1,108 @@
+//! [`CallbackClient`] definitions.
+//!
+//! [`CallbackClient`]: Client
+
+use async_trait::async_trait;
+use derive_more::From;
+use time::OffsetDateTime;
+
+use crate::{ErrorResponse, StatefulFid};
+
+/// Abstraction of a [Control API] callback client.
+///
+/// [Control API]: https://tinyurl.com/yxsqplq7
+#[async_trait(?Send)]
+pub trait Client {
+    /// Sends provided [`Request`].
+    async fn send(&self, request: Request) -> Result<(), ErrorResponse>;
+}
+
+/// [Control API] callback.
+///
+/// Used for sending callbacks with [`Client::send`].
+///
+/// [Control API]: https://tinyurl.com/yxsqplq7
+#[derive(Debug)]
+pub struct Request {
+    /// `FID` (Full `ID`) of element with which event was occurred.
+    pub fid: StatefulFid,
+
+    /// [`Member::on_join`] or [`Member::on_leave`] callback `URL`.
+    ///
+    /// [`Member::on_join`]: crate::Member::on_join
+    /// [`Member::on_leave`]: crate::Member::on_leave
+    pub url: String,
```

Comment:
@ilslv why do we have `url` here? It should be unnecessary.

Reply:
@ilslv the `CallbackClient` represents a client for calling that URL. So it should be something like:
```rust
let callback = grpc::CallbackClient::new("grpc://127.0.0.1/on_join");
let resp = callback.send(Request { fid: "room/1".into(), at: DateTime::now_utc(), event: OnJoinEvent.into() }).await;
```

Reply:
@tyranron why should we limit ourself to a single `url`? Why just don't store connection pool inside `grpc::CallbackClient`?

```rust
let callback = grpc::CallbackClient::new();
let resp = callback.send(Request { 
    fid: "room/1".into(), 
    url: "grpc://127.0.0.1/on_join", // Creates a new connection or reuses in case it's already present
    at: DateTime::now_utc(), 
    event: OnJoinEvent.into() 
}).await;
```

And what to do in case there are multiple callback `url`s? Create a bunch of them with fixed known `url`s  and then error if callback server changes?

Reply:
@ilslv 

> why should we limit ourself to a single `url`? Why just don't store connection pool inside `grpc::CallbackClient`?

Because from the API point of view there is no sense for a single callback to operate on multiple URLs. What we want is just to trasform the provided callback URI into a callback object (create the callback) and then just call this object many times later (fire the callback). Mixing creation and firing phases makes the API not quite clear.



> And what to do in case there are multiple callback `url`s? Create a bunch of them with fixed known `url`s 

Yes. We create a callback object from its URI representation.

>  and then error if callback server changes?

If we change the callback server, the it means we have different callback's URI, so this is a totally different callback object. We shoud create a new callback object, not error on the old one.



> And why do we have `on_join` and `on_leave` in `Member`s spec in a first place?

Because this informs what actually has happened on a media server.

Such scheme allows both to use the same URL for different callbacks and to use a separate URL for each callback. Integrator chooses what is better suits his needs. For example, Kafka sink will represent a single endpoint just relaying all the events directly into a Kafka topic. Some HTTP integration may be easier to have callbacks already routed by URL path, so multiple URLs may be used.

> ```rust
>     url: "grpc://127.0.0.1/on_join", // Creates a new connection or reuses in case it's already present
> ```

Nothing prevents us from reusing connections in different callback objects pointing to the same host. It's just an implementation detail.

Reply:
@tyranron so, from my understanding, `CallbackClient` will have 2 methods:

```rust
```

Reply:
@tyranron so, from my understanding, `CallbackClient` will have 2 methods:

```rust
#[async_trait(?Send)]
pub trait Client {
    async fn new(url: String) -> Self;    

    async fn send(&self, request: Request) -> Result<(), ErrorResponse>;
    //              No `url` here ^^^^^^^
}
```

So just merging `Factory` and `CallbackClient`.

Reply:
@ilslv no, it's better to keep them as different `trait`s, because factory and callback itself may be different types. Also, it's nice to have `Client` being object-safe, while the `Factory` will definitely disallow this.

Reply:
@tyranron to make sure, we are back to exposing `CallbackClientFactory` and `CallbakcClient` in a public API?

Reply:
@ilslv `ClientFactory` is not a necessary abstraction. It doesn't interact with any other library parts and is solely a media server implementation detail, still not mandatory one. But we could provide it here as the "included battery".